drivers/frequency/adf4382/adf4382.c

```diff
@@ -0,0 +1,1199 @@
+ */
+int adf4382_spi_write(struct adf4382_dev *dev, uint16_t reg_addr, uint8_t data)
+{
+	uint8_t buff[ADF4382_BUFF_SIZE_BYTES];
+	uint16_t cmd;
+
+	cmd = ADF4382_SPI_WRITE_CMD | reg_addr;
+	if (dev->spi_desc->bit_order) {
+		buff[0] = no_os_bit_swap_constant_8(cmd & 0xFF);
+		buff[1] = no_os_bit_swap_constant_8(cmd >> 8);
+		buff[2] = no_os_bit_swap_constant_8(data);
+	} else {
+		buff[0] = cmd >> 8;
+		buff[1] = cmd & 0xFF;
+		buff[2] = data;
+	}
+
+	return no_os_spi_write_and_read(dev->spi_desc, buff,
+					ADF4382_BUFF_SIZE_BYTES);
+}
+
+/**
+ * @brief Reads data from ADF4382 over SPI.
+ * @param dev 	   - The device structure.
+ * @param reg_addr - The register address.
+ * @param data	   - Data read from the device.
+ * @return	   - 0 in case of success or negative error code otherwise.
+ */
+int adf4382_spi_read(struct adf4382_dev *dev, uint16_t reg_addr, uint8_t *data)
+{
+	int ret;
+	uint8_t buff[ADF4382_BUFF_SIZE_BYTES];
+	uint16_t cmd;
+
+	cmd = ADF4382_SPI_READ_CMD | reg_addr;
+	if (dev->spi_desc->bit_order) {
+		buff[0] = no_os_bit_swap_constant_8(cmd & 0xFF);
+		buff[1] = no_os_bit_swap_constant_8(cmd >> 8);
+		buff[2] = no_os_bit_swap_constant_8(ADF4382_SPI_DUMMY_DATA);
+	} else {
+		buff[0] = cmd >> 8;
+		buff[1] = cmd & 0xFF;
+		buff[2] = ADF4382_SPI_DUMMY_DATA;
+	}
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buff,
+				       ADF4382_BUFF_SIZE_BYTES);
+	if(ret < 0)
+		return ret;
+
+	*data = buff[2];
+
+	return ret;
+}
+
+/**
+ * @brief Updates the values of the ADF4382 register.
+ * @param dev 	   - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask 	   - Bits to be updated.
+ * @param data 	   - Update value for the mask.
+ * @return	   - 0 in case of success or negative error code otherwise.
+ */
+int adf4382_spi_update_bits(struct adf4382_dev *dev, uint16_t reg_addr,
+			    uint8_t mask, uint8_t data)
+{
+	uint8_t tmp, orig;
+	int ret;
+
+	ret = adf4382_spi_read(dev, reg_addr, &orig);
+	if (ret)
+		return ret;
+
+	tmp = orig & ~mask;
+	tmp |= data & mask;
+
+	if (tmp != orig)
+		return adf4382_spi_write(dev, reg_addr, tmp);
+
+	return 0;
+}
+
+/**
+ * @brief Tests if bits are set in the ADF4382 register.
+ * @param dev 	   - The device structure.
+ * @param reg_addr - The register address.
+ * @param bits 	   - Bits to be tested.
+ * @return	   - 0 in case bits are not set, positive value if they are set
+ * 		     or negative error code otherwise.
+ */
+int adf4382_spi_test_bits(struct adf4382_dev *dev, uint16_t reg_addr,
+			  uint8_t bits)
+{
+	uint8_t val, ret;
+
+	ret = adf4382_spi_read(dev, reg_addr, &val);
+	if (ret)
+		return ret;
+
+	return (val & bits) == bits;
+}
+
+/**
+ * @brief Will output on the terminal the values of all the ADF4382 registers.
+ * @param dev 	- The device structure.
+ * @return 	- 0 in case of success or negative error code.
+ */
+int adf4382_reg_dump(struct adf4382_dev *dev)
+{
+	uint8_t val;
+	uint8_t ret;
+	uint16_t i;
+
+	for(i = 0; i < 0x68; i++) {
```

Comment:
usually we use macro definitions for the registers. but this is probably similar to adf43777 and the register naming is actually the value of the register. looks good like this or macro `AD4382_REG(x)` can be used. Let's see what others prefer.

Reply:
This is a debugging only function that users wouldn't not use in "production". I wouldn't care much about that. It's probably even easier to follow the datasheet as Cipi wrote it because you normally have a large table with register names and addresses and you see gaps/discontinuities in addresses right away when you look at the addresses.